doc/man/gnunet-scalarproduct.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd August 8, 2013
.Dt GNUNET-SCALARPRODUCT 1
.Os
.Sh NAME
.Nm gnunet-scalarproduct
.Nd compute a vectorproduct
.Sh SYNOPSIS
.Nm
.Op Fl e Ar ELEMENTS | Fl -elements= Ns Ar ELEMENTS
.Op Fl m Ar MASK | Fl -mask= Ns Ar MASK
.Op Fl k Ar KEY | Fl -key= Ns Ar KEY
.Op Fl c Ar FILENAME | -config= Ns Ar FILENAME
.Op Fl p Ar PEERID | Fl -peer= Ns Ar PEERID
.Op Fl h | -help
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl v | -version
.Sh DESCRIPTION
.Nm
enables you to compute a vectorproduct across two peers
.Sy Alice
and
.Sy Bob .
.Pp
A client can issue one of two messages to its service:
.Bl -enum -width 3n -offset indent
.It
A request to compute a vectorproduct with another peer (Alice)
.It
Elements to support a peer in computing a vectorproduct (Bob)
.El
.Pp
Both requests must share the same SID, which can be an arbitrary
string identifying the session. SIDs should be unique, however it is
sufficient to guarantee the uniqueness of the tupel element count and
session ID.
.Pp
Alice's client must supply the ASCII encoded peer ID of bob's
service, it will internally be checked by the client for
validity. Invalid values here result in the client or the service
failing the session.
.Pp
Elements are handed over as signed decimal integers, the element count
supplied by Alice and Bob must match. Alice can also
supply a mask for these values to her service, which allows partial
vector products to be computed across the vector. Elements can be
masked by setting their the corresponding mask element to zero, any
other value means the element will not be masked. Alice's
client will also mask all 0-values to avoid information leakage to
Bob.
.Pp
The protocol by definition relies on Alice and Bob being
benign, thus Bob can arbitrarily falsify his information. Both
peers collaborate to achieve a correct result.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl e Ar ELEMENTS | Fl -elements= Ns Ar ELEMENTS
The element-vector the vectorproduct should be computed over in
signed decimal form, eg: "42,1,-3,3,7".
Zero value elements will be automatically masked.
.It Fl m Ar MASK | Fl -mask= Ns Ar MASK
Elements in the vector can be masked.
There must be at least two elements left in the vector to
compute a vectorproduct.
Non-Zero values indicate an element is not maskes.
.It Fl k Ar KEY | Fl -key= Ns Ar KEY
The session key, a shared string of arbitrary length from which
the SID will be generated.
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Use the configuration file FILENAME.
.It Fl p Ar PEERID | Fl -peer= Ns Ar PEERID
The remote peer's ASCII-armored gnunet-peer ID as output by
.Xr gnunet-peerinfo 1 .
If this option is not given, the peer will take the Bob's role.
.It Fl h | -help
Print short help on options.
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Use LOGLEVEL for logging.
Valid values are DEBUG, INFO, WARNING and ERROR.
.It Fl v | -version
Print GNUnet version number.
.El
.Sh SEE ALSO
.Xr gnunet-peerinfo 1
.sp
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


ChangeLog
Wed May 29 11:17:49 2019 +0200
  Added CAA record type (RFC 6844). -schanzen

Sat 25 May 2019 08:27:39 AM CEST
  Significantly speed-up ego creation and deletion via
  gnunet-identity command when there are many (hundreds)
  of egos at the same peer. -CG

Sun May 12 12:00:00 2019 +0200
  Releasing GNUnet 0.11.4. -MS

Thu May 9 23:35:24 2019 +0200
  gnunet-identity can now output the public key on creation of new
  egos, and -e can act as a filter when listing all egos. -CG

Thu May 2 08:44:37 2019 +0200
  gnunet-arm now logs to file by default -CG

Sat Apr 27 21:50:57 2019 +0200
  RECLAIM: Added ECDH encryption for authorization code payload -schanzen

Sat 27 Apr 2019 06:36:31 AM CEST
  Bump libmicrohttpd version requirement as some intermediate
  versions have relevant bugs that cause problems. -CG

Sat Apr 27 01:50:43 2019 +0200
  RECLAIM: No longer experimental -schanzen

Fri Apr 26 11:08:21 2019 +0200
  REST: Configuration plugin added -schanzen

Tue Apr 23 23:00:00 UTC 2019
  All manpages are properly licensed. -ng0

Sun Apr 21 22:22:22 UTC 2019
  All manpages are now in mdoc format. -ng0

Sat Apr 20 18:38:43 UTC 2019
  Remove optional gnunet-download-manager.scm and
  with it the optional dependency on a no longer
  maintained version of GNU Guile (#5682). -ng0

Sun Apr 14 09:54:20 2019 +0200
  RECLAIM: Moved away from ABE to GNS encryption. Minor API changed and
  renaming of records. Deprecated SQLite plugin in favor of namestore.
  -schanzen

Sun Apr 7 14:40:59 2019 +0200
  Releasing GNUnet 0.11.3. -CG

Sat 06 Apr 2019 02:36:17 PM CEST
  Introducing GNUNET_memcmp() and GNUNET_is_zero() for better typing
  when comparing memory areas. -CG

Fri 05 Apr 2019 11:38:53 PM CEST
  Use paths of sysctl, ip, ifconfig and ip(6)tables from configure
  if they work. (#5352). -CG

Thu 04 Apr 2019 12:39:44 PM CEST
  Hotfix for build issue. Releasing GNUnet 0.11.2. -CG/MS

Wed 03 Apr 2019 03:18:43 PM CEST
  Releasing GNUnet 0.11.1. -CG

Wed 03 Apr 2019 01:54:44 PM CEST
  Merged gnunet-qr-c-2 branch, replacing Python-based implementation
  of gnunet-qr with C-based implementation that also integrates
  gnunet-uri. -HG/MS/CG

Wed 03 Apr 2019 01:15:57 PM CEST
  Fixed a few test case regressions (GNS, FS). -CG

Mon Apr 1 22:00:00 UTC 2019
  documentation/buildsystem: manpages and other documentation parts
  can now be build more controlled again, this fixes the regression
  in 0.11.0 where documentation was an on/off switch.
  New introduced: --enable-include-manpages.
  Fixed: texi2mdoc integration (--enable-texi2mdoc-generation).
  buildsystem: decrease python dependency to 3.4 from
  previously 3.7.
  -ng0

Tue Mar 12 10:06:17 CET 2019
  REST: Move back from rest-plugin into respective subsystem directories.

Mon Mar 11 02:47:55 CET 2019
  Make configure complain if no working libidn is detected. -CG

Sat Mar 9 15:58:45 2019 +0100
  REST: Config option for address bind. Defaults to localhost.
  GNS: dns2gns/gns-proxy config option for address bind. Defaults to localhost.

Sat Mar 9 01:58:22 CET 2019
  gnunet-publish now by default does not expose the creation time,
  option -d no longer exists, to enable the old behavior use -E. -CG

Sat Mar 2 14:23:10 2019 +0100
  utils: gnunet-qr.py now installed under pkgshare instead of bin

Thu Feb 28 19:07:59 CET 2019
  CADET: Fix in order, unrealiable message ordering

Thu 28 Feb 2019 10:54:58 AM CET
  Releasing GNUnet 0.11.0. -CG

Wed Jun  6 08:28:27 CEST 2018
  Releasing GNUnet 0.11.0pre66. Let's try to keep a better
  ChangeLog again henceforth. -CG

Sat Jan  7 16:41:34 CET 2017
  Converting NAT library to new NAT service (and
  splitting of nat-auto service for auto-configuration). -CG


src/scalarproduct/gnunet-scalarproduct.c
/*
     This file is part of GNUnet.
     Copyright (C) 2013, 2014 GNUnet e.V.

     GNUnet is free software: you can redistribute it and/or modify it
     under the terms of the GNU Affero General Public License as published
     by the Free Software Foundation, either version 3 of the License,
     or (at your option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Affero General Public License for more details.
    
     You should have received a copy of the GNU Affero General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */

/**
 * @file scalarproduct/gnunet-scalarproduct.c
 * @brief scalarproduct client
 * @author Christian M. Fuchs
 */
#define GCRYPT_NO_DEPRECATED
#include <gcrypt.h>
#include <inttypes.h>

#include "platform.h"
#include "gnunet_util_lib.h"
#include "gnunet_scalarproduct_service.h"
#include "gnunet_protocols.h"
#include "scalarproduct.h"

#define LOG(kind,...) GNUNET_log_from (kind, "gnunet-scalarproduct",__VA_ARGS__)


/**
 * the session key identifying this computation
 */
static struct GNUNET_HashCode session_key;

/**
 * PeerID we want to compute a scalar product with
 */
static struct GNUNET_PeerIdentity peer_id;

/**
 * Option -p: destination peer identity for checking message-ids with
 */
static char *input_peer_id;

/**
 * Option -p: destination peer identity for checking message-ids with
 */
static char *input_session_key;

/**
 * Option -e: vector to calculate a scalarproduct with
 */
static char *input_elements;

/**
 * Global return value
 */
static int ret = -1;

/**
 * our Scalarproduct Computation handle
 */
static struct GNUNET_SCALARPRODUCT_ComputationHandle *computation;


/**
 * Callback called if we are initiating a new computation session
 *
 * @param cls unused
 * @param status if our job was successfully processed
 */
static void
responder_callback (void *cls,
                    enum GNUNET_SCALARPRODUCT_ResponseStatus status)
{
  switch (status)
  {
  case GNUNET_SCALARPRODUCT_STATUS_SUCCESS:
    ret = 0;
    LOG (GNUNET_ERROR_TYPE_INFO,
         "Session %s concluded.\n",
         GNUNET_h2s (&session_key));
    break;
  case GNUNET_SCALARPRODUCT_STATUS_INVALID_RESPONSE:
    LOG (GNUNET_ERROR_TYPE_ERROR,
         "Session %s failed: invalid response\n",
         GNUNET_h2s (&session_key));
    break;
  case GNUNET_SCALARPRODUCT_STATUS_FAILURE:
    LOG (GNUNET_ERROR_TYPE_ERROR,
         "Session %s failed: service failure\n",
         GNUNET_h2s (&session_key));
    break;
  case GNUNET_SCALARPRODUCT_STATUS_DISCONNECTED:
    LOG (GNUNET_ERROR_TYPE_ERROR,
         "Session %s failed: service disconnect!\n",
         GNUNET_h2s (&session_key));
    break;
  default:
    LOG (GNUNET_ERROR_TYPE_ERROR,
         "Session %s failed: return code %d\n",
         GNUNET_h2s (&session_key),
         status);
  }
  computation = NULL;
  GNUNET_SCHEDULER_shutdown ();
}


/**
 * Callback called if we are initiating a new computation session
 *
 * @param cls unused
 * @param status if our job was successfully processed
 * @param result the result in gnu/gcry MPI format
 */
static void
requester_callback (void *cls,
                    enum GNUNET_SCALARPRODUCT_ResponseStatus status,
                    gcry_mpi_t result)
{
  unsigned char *buf;
  gcry_error_t rc;

  switch (status)
  {
  case GNUNET_SCALARPRODUCT_STATUS_SUCCESS:
    if (0 == (rc = gcry_mpi_aprint (GCRYMPI_FMT_HEX, &buf, NULL, result)))
    {
      ret = 0;
      fprintf (stdout,
               "%s\n",
               buf);
      fflush (stdout);
    }
    else
      LOG_GCRY (GNUNET_ERROR_TYPE_ERROR,
                "gcry_mpi_aprint",
                rc);
    break;
  case GNUNET_SCALARPRODUCT_STATUS_INVALID_RESPONSE:
    LOG (GNUNET_ERROR_TYPE_ERROR,
         "Session %s with peer %s failed: invalid response received\n",
         GNUNET_h2s (&session_key),
         GNUNET_i2s (&peer_id));
    break;
  case GNUNET_SCALARPRODUCT_STATUS_FAILURE:
    LOG (GNUNET_ERROR_TYPE_ERROR,
         "Session %s with peer %s failed: API failure\n",
         GNUNET_h2s (&session_key),
         GNUNET_i2s (&peer_id));
    break;
  case GNUNET_SCALARPRODUCT_STATUS_DISCONNECTED:
    LOG (GNUNET_ERROR_TYPE_ERROR,
         "Session %s with peer %s was disconnected from service.\n",
         GNUNET_h2s (&session_key),
         GNUNET_i2s (&peer_id));
    break;
  default:
    LOG (GNUNET_ERROR_TYPE_ERROR,
         "Session %s with peer %s failed: return code %d\n",
         GNUNET_h2s (&session_key),
         GNUNET_i2s (&peer_id),
         status);
  }
  computation = NULL;
  GNUNET_SCHEDULER_shutdown ();
}


/**
 * Task run during shutdown.
 *
 * @param cls unused
 * @param tc unused
 */
static void
shutdown_task (void *cls)
{
  if (NULL != computation)
  {
    GNUNET_SCALARPRODUCT_cancel (computation);
    ret = 1; /* aborted */
  }
}


/**
 * Main function that will be run by the scheduler.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  char *begin = input_elements;
  char *end;
  unsigned int i;
  struct GNUNET_SCALARPRODUCT_Element *elements;
  uint32_t element_count = 0;

  if (NULL == input_elements)
  {
    LOG (GNUNET_ERROR_TYPE_ERROR,
         _("You must specify at least one message ID to check!\n"));
    return;
  }
  if ( (NULL == input_session_key) ||
       (0 == strlen (input_session_key)) )
  {
    LOG (GNUNET_ERROR_TYPE_ERROR,
         _("This program needs a session identifier for comparing vectors.\n"));
    return;
  }
  GNUNET_CRYPTO_hash (input_session_key,
                      strlen (input_session_key),
                      &session_key);
  if ( (NULL != input_peer_id) &&
       (GNUNET_OK !=
        GNUNET_CRYPTO_eddsa_public_key_from_string (input_peer_id,
                                                    strlen (input_peer_id),
                                                    &peer_id.public_key)) )
  {
    LOG (GNUNET_ERROR_TYPE_ERROR,
         _("Tried to set initiator mode, as peer ID was given. "
           "However, `%s' is not a valid peer identifier.\n"),
         input_peer_id);
    return;
  }
  if ( ('\'' == *begin) &&
       ('\'' == begin[strlen(begin)-1]) )
  {
    begin[strlen(begin)-1] = '\0';
    if (strlen (begin) > 0)
      begin++;
  }
  for (end = begin; 0 != *end; end++)
    if (*end == ';')
      element_count++;
  if (0 == element_count)
  {
    LOG (GNUNET_ERROR_TYPE_ERROR,
         _("Need elements to compute the scalarproduct, got none.\n"));
    return;
  }

  elements = GNUNET_malloc (sizeof(struct GNUNET_SCALARPRODUCT_Element) * element_count);

  for (i = 0; i < element_count;i++)
  {
    struct GNUNET_SCALARPRODUCT_Element element;
    char* separator = NULL;

    /* get the length of the current key,value; tupel */
    for (end = begin; *end != ';'; end++)
      if (*end == ',')
        separator = end;

    /* final element */
    if ( (NULL == separator) ||
         (begin == separator) ||
         (separator == end - 1) )
    {
      LOG (GNUNET_ERROR_TYPE_ERROR,
           _("Malformed input, could not parse `%s'\n"),
           begin);
      GNUNET_free (elements);
      return;
    }
    *separator = 0;
    /* read the element's key */
    GNUNET_CRYPTO_hash (begin,
                        strlen (begin),
                        &element.key);

    /* read the element's value */
    if (1 !=
        sscanf (separator + 1,
                "%" SCNd64 ";",
                &element.value) )
    {
      LOG (GNUNET_ERROR_TYPE_ERROR,
           _("Could not convert `%s' to int64_t.\n"),
           begin);
      GNUNET_free (elements);
      return;
    }
    element.value = GNUNET_htonll (element.value);
    elements[i] = element;
    begin = end + 1;
  }

  if ( ( (NULL != input_peer_id) &&
         (NULL == (computation
                   = GNUNET_SCALARPRODUCT_start_computation (cfg,
                                                             &session_key,
                                                             &peer_id,
                                                             elements, element_count,
                                                             &requester_callback,
                                                             NULL))) ) ||
       ( (NULL == input_peer_id) &&
         (NULL == (computation
                   = GNUNET_SCALARPRODUCT_accept_computation (cfg,
                                                              &session_key,
                                                              elements, element_count,
                                                              &responder_callback,
                                                              NULL))) ) )
  {
    fprintf (stderr,
             _("Failed to initiate computation, were all keys unique?\n"));
    GNUNET_free (elements);
    return;
  }
  GNUNET_free (elements);
  GNUNET_SCHEDULER_add_shutdown (&shutdown_task,
				 NULL);
  ret = 0;
}


/**
 * The main function to the scalarproduct client.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_GETOPT_CommandLineOption options[] = {

    GNUNET_GETOPT_option_string ('e',
                                 "elements",
                                 "\"key1,val1;key2,val2;...,keyn,valn;\"",
                                 gettext_noop ("A comma separated list of elements to compare as vector with our remote peer."),
                                 &input_elements),

    GNUNET_GETOPT_option_string ('e',
                                 "elements",
                                 "\"key1,val1;key2,val2;...,keyn,valn;\"",
                                 gettext_noop ("A comma separated list of elements to compare as vector with our remote peer."),
                                 &input_elements),

    GNUNET_GETOPT_option_string ('p',
                                 "peer",
                                 "PEERID",
                                 gettext_noop ("[Optional] peer to calculate our scalarproduct with. If this parameter is not given, the service will wait for a remote peer to compute the request."),
                                 &input_peer_id),

    GNUNET_GETOPT_option_string ('k',
                                 "key",
                                 "TRANSACTION_ID",
                                 gettext_noop ("Transaction ID shared with peer."),
                                 &input_session_key),

    GNUNET_GETOPT_OPTION_END
  };

  return (GNUNET_OK ==
          GNUNET_PROGRAM_run (argc,
                              argv,
                              "gnunet-scalarproduct",
                              gettext_noop ("Calculate the Vectorproduct with a GNUnet peer."),
                              options, &run, NULL)) ? ret : 1;
}

/* end of gnunet-scalarproduct.c */
